qemu: upgrade to 8.2.4

ryansavino opened the pull request:
There is a known issue in qemu 7.2.0 that causes kernel-hashes to fail the verification of the launch binaries for the SEV legacy use case.

Fixes: #9148

niteeshkd commented:
@ryansavino 
I think it would need the file `kata-containers/qemu/patches/tag_patches/v8.2.1/no_patches.txt` otherwise kata-deploy scripts will complain.

stevenhorsman commented:
/test

wainersm commented:
Hi @ryansavino !

Please, refresh my mind: sev relies on vanilla QEMU, whereas SNP on qemu-snp (a fork of upstream qemu). Am I right?

Also, I would like to check the relation of this bump with https://github.com/kata-containers/kata-containers/pull/9174 . Does 9174 superseeds this one or they aren't related at all?

ryansavino commented:
> Hi @ryansavino !
> 
> Please, refresh my mind: sev relies on vanilla QEMU, whereas SNP on qemu-snp (a fork of upstream qemu). Am I right?
> 
> Also, I would like to check the relation of this bump with #9174 . Does 9174 superseeds this one or they aren't related at all?

Correct, the new 8.2.1 vanilla qemu for SEV legacy. For SNP 9174 will fix the snp experimental qemu to work with the latest host kernel (host kernels already upgraded on amd kata ci nodes). They are separate, but the SNP test will fail here most likely until 9174 is merged.

ryansavino commented:
> jenkins-ci-ARM-ubuntu-20.04-main

Added, thanks!

ryansavino commented:
> @ryansavino I think it would need the file `kata-containers/qemu/patches/tag_patches/v8.2.1/no_patches.txt` otherwise kata-deploy scripts will complain.

Added, thanks!

ChengyuZhu6 reviewed (approved):
LGTM thanks @ryansavino

stevenhorsman commented:
Hi Ryan, we discussed this on the AC meeting and @amshinde mentioned that there are typically some metrics and feature check done when upgrading QEMU, so she is going to give some guidance here. In the meantime I'll mark this as `do-not-merge` to make sure she has the chance to share them.

amshinde commented:
@ryansavino Please take a look at this script : https://github.com/kata-containers/kata-containers/blob/main/tools/packaging/scripts/configure-hypervisor.sh.
It is used to disable any features that may be enabled by default with upgrading to the next major version in order to not have features not required by Kata enabled by default. This is done to make sure we have a minimal memory footprint important from performance and security point of view.
In addition the script is also used to enable any qemu features that may helo improve performance.

ryansavino commented:
I've changed this to 8.2.4 qemu; tests are passing. I've done an analysis on the enabled/disabled build command line options. Here's what I've found.

New flags in 8.2.4 (in reference to 7.2.0):
```
>     --disable-af-xdp) printf "%s" -Daf_xdp=disabled ;;
>     --disable-avx512bw) printf "%s" -Davx512bw=disabled ;;
>     --disable-colo-proxy) printf "%s" -Dcolo_proxy=disabled ;;
>     --disable-debug-info) printf "%s" -Ddebug=false ;;
>     --disable-debug-graph-lock) printf "%s" -Ddebug_graph_lock=false ;;
>     --disable-debug-tcg) printf "%s" -Ddebug_tcg=false ;;
>     --disable-hexagon-idef-parser) printf "%s" -Dhexagon_idef_parser=false ;;
>     --disable-hv-balloon) printf "%s" -Dhv_balloon=disabled ;;
>     --disable-libdw) printf "%s" -Dlibdw=disabled ;;
>     --disable-libkeyutils) printf "%s" -Dlibkeyutils=disabled ;;
>     --disable-modules) printf "%s" -Dmodules=disabled ;;
>     --disable-pipewire) printf "%s" -Dpipewire=disabled ;;
>     --disable-pixman) printf "%s" -Dpixman=disabled ;;
>     --disable-plugins) printf "%s" -Dplugins=false ;;
>     --disable-relocatable) printf "%s" -Drelocatable=false ;;
>     --disable-rutabaga-gfx) printf "%s" -Drutabaga_gfx=disabled ;;
>     --disable-safe-stack) printf "%s" -Dsafe_stack=false ;;
>     --disable-sanitizers) printf "%s" -Dsanitizers=false ;;
>     --disable-stack-protector) printf "%s" -Dstack_protector=disabled ;;
>     --disable-tsan) printf "%s" -Dtsan=false ;;
>     --disable-vhdx) printf "%s" -Dvhdx=disabled ;;
>     --disable-virtfs-proxy-helper) printf "%s" -Dvirtfs_proxy_helper=disabled ;;
>     --disable-vmdk) printf "%s" -Dvmdk=disabled ;;
>     --disable-vpc) printf "%s" -Dvpc=disabled ;;
>     --disable-werror) printf "%s" -Dwerror=false ;;
```

Removed flags in 8.2.4 (in reference to 7.2.0):
```
<     --disable-gprof) printf "%s" -Dgprof=false ;;
<     --disable-hax) printf "%s" -Dhax=disabled ;;
<     --disable-profiler) printf "%s" -Dprofiler=false ;;
<     --disable-virtiofsd) printf "%s" -Dvirtiofsd=disabled ;;
```

Same exact findings as above for the `--enable` flags between 7.2.0 and 8.2.4.

I'm not a qemu feature expert here. Hoping someone can help me decipher if any of these should be enabled or disabled in the `configure-hypervisor.sh`. @amshinde @gkurz @fidencio @wainersm

fidencio commented:
@ryansavino, not exactly providing you an answer, but rather providing you data that can be used for now and future bumps.

I like the approach @danilocesar took at Fedora / RHEL builds of QEMU a few years ago (and it's still used).
Here you can see that there, they disable everything: https://src.fedoraproject.org/rpms/qemu/blob/rawhide/f/qemu.spec#_1530

And then later on they come back enabling only what they want to support:
https://src.fedoraproject.org/rpms/qemu/blob/rawhide/f/qemu.spec#_1716

This is an approach that I'd love to see taken as part of our builds.  Is this needed for your PR, though?  I'd say no, although the improvement would be more than welcome.

ryansavino commented:
From an initial investigation, I'm going to try testing with several options disabled that seem reasonable:
```
qemu_options+=(size:--disable-colo-proxy)
qemu_options+=(size:--disable-debug-info)
qemu_options+=(size:--disable-debug-graph-lock)
qemu_options+=(size:--disable-hexagon-idef-parser)
qemu_options+=(size:--disable-libdw)
qemu_options+=(size:--disable-modules)
qemu_options+=(size:--disable-pipewire)
qemu_options+=(size:--disable-pixman)
qemu_options+=(size:--disable-plugins)
qemu_options+=(size:--disable-relocatable)
qemu_options+=(size:--disable-rutabaga-gfx)
qemu_options+=(size:--disable-stack-protector)
qemu_options+=(size:--disable-vmdk)
```

Here are some options that I wasn't sure about after looking them up:
```
# 8.2.4: Potential optimizations to enable
qemu_options+=(size:--enable-af-xdp)      # Networking stack
qemu_options+=(size:--enable-avx512bw)    # Intel instruction set
qemu_options+=(size:--enable-hv-balloon)  # HyperV memory ballooning
qemu_options+=(size:--enable-libkeyutils) # kernel based keystore
qemu_options+=(size:--enable-safe-stack)  # Code-Pointer integrity
qemu_options+=(size:--enable-sanitizers)  # memory bug detection
qemu_options+=(size:--enable-tsan)        # thread sanitizer
qemu_options+=(size:--enable-vhdx)        # HyperV virtual disk
qemu_options+=(size:--enable-vpc)         # IBM cloud image
qemu_options+=(size:--disable-werror)     # Do not die on warnings
```

Would help if stakeholders could provide some suggestion here.

@stevenhorsman @BbolroC - vpc
@fidencio - avx512bw
@danmihai1 @jepio - hv-balloon, vhdx
@wainersm @amshinde @gkurz @sameo @fitzthum

jepio commented:
> @danmihai1 @jepio - hv-balloon, vhdx

These would only be useful for windows. hv-balloon for windows guests, vhdx for windows hosts. Safe to say "disable" from me, unless @fidencio knows something i don't.

fitzthum commented:
It's tough to be super thorough here, but I think the things you are disabling seem fine and I wouldn't bother with any of the enable flags.

stevenhorsman commented:
@gkurz @amshinde - you probably have more idea about this than I do, so can you take a look at Ryan's list of options? Thanks!

gkurz commented:
> I've changed this to 8.2.4 qemu; tests are passing. I've done an analysis on the enabled/disabled build command line options. Here's what I've found.
> 
> New flags in 8.2.4 (in reference to 7.2.0):
> 
> ```
> >     --disable-af-xdp) printf "%s" -Daf_xdp=disabled ;;
> >     --disable-avx512bw) printf "%s" -Davx512bw=disabled ;;
> >     --disable-colo-proxy) printf "%s" -Dcolo_proxy=disabled ;;
> >     --disable-debug-info) printf "%s" -Ddebug=false ;;

This flag is definitely not new. 

```
$ git grep debug-info v7.2.0 -- configure
v7.2.0:configure:  --enable-debug-info) meson_option_add -Ddebug=true
v7.2.0:configure:  --disable-debug-info) meson_option_add -Ddebug=false
v7.2.0:configure:  debug-info      debugging information
```

It was moved to meson by https://gitlab.com/qemu-project/qemu/-/commit/c0e705c6800a1f98c8cb3795ce66f5a48d4e33aa during the v8.1.0 development.

Similar remark for some other options in this list.

> >     --disable-debug-graph-lock) printf "%s" -Ddebug_graph_lock=false ;;
> >     --disable-debug-tcg) printf "%s" -Ddebug_tcg=false ;;
> >     --disable-hexagon-idef-parser) printf "%s" -Dhexagon_idef_parser=false ;;
> >     --disable-hv-balloon) printf "%s" -Dhv_balloon=disabled ;;
> >     --disable-libdw) printf "%s" -Dlibdw=disabled ;;
> >     --disable-libkeyutils) printf "%s" -Dlibkeyutils=disabled ;;
> >     --disable-modules) printf "%s" -Dmodules=disabled ;;
> >     --disable-pipewire) printf "%s" -Dpipewire=disabled ;;
> >     --disable-pixman) printf "%s" -Dpixman=disabled ;;
> >     --disable-plugins) printf "%s" -Dplugins=false ;;
> >     --disable-relocatable) printf "%s" -Drelocatable=false ;;
> >     --disable-rutabaga-gfx) printf "%s" -Drutabaga_gfx=disabled ;;
> >     --disable-safe-stack) printf "%s" -Dsafe_stack=false ;;
> >     --disable-sanitizers) printf "%s" -Dsanitizers=false ;;
> >     --disable-stack-protector) printf "%s" -Dstack_protector=disabled ;;
> >     --disable-tsan) printf "%s" -Dtsan=false ;;
> >     --disable-vhdx) printf "%s" -Dvhdx=disabled ;;
> >     --disable-virtfs-proxy-helper) printf "%s" -Dvirtfs_proxy_helper=disabled ;;
> >     --disable-vmdk) printf "%s" -Dvmdk=disabled ;;
> >     --disable-vpc) printf "%s" -Dvpc=disabled ;;
> >     --disable-werror) printf "%s" -Dwerror=false ;;
> ```
> 
> Removed flags in 8.2.4 (in reference to 7.2.0):
> 
> ```
> <     --disable-gprof) printf "%s" -Dgprof=false ;;
> <     --disable-hax) printf "%s" -Dhax=disabled ;;
> <     --disable-profiler) printf "%s" -Dprofiler=false ;;
> <     --disable-virtiofsd) printf "%s" -Dvirtiofsd=disabled ;;
> ```
> 
> Same exact findings as above for the `--enable` flags between 7.2.0 and 8.2.4.
> 

`scripts/meson-buildoptions.sh` isn't enough. `configure` must be considered as well.

> I'm not a qemu feature expert here. Hoping someone can help me decipher if any of these should be enabled or disabled in the `configure-hypervisor.sh`. @amshinde @gkurz @fidencio @wainersm

ryansavino commented:
> `scripts/meson-buildoptions.sh` isn't enough. `configure` must be considered as well.

Ah, I was only checking the meson build. I've reanalyzed considering the `configure` as well and removed these options from the list:
```
qemu_options+=(size:--disable-debug-info)
qemu_options+=(size:--disable-modules)
qemu_options+=(size:--disable-plugins)
qemu_options+=(size:--disable-stack-protector)
```

New list of disabled features:
```
qemu_options+=(size:--disable-colo-proxy)
qemu_options+=(size:--disable-debug-graph-lock)
qemu_options+=(size:--disable-hexagon-idef-parser)
qemu_options+=(size:--disable-libdw)
qemu_options+=(size:--disable-pipewire)
qemu_options+=(size:--disable-pixman)
qemu_options+=(size:--disable-relocatable)
qemu_options+=(size:--disable-rutabaga-gfx)
qemu_options+=(size:--disable-vmdk)
qemu_options+=(size:--disable-avx512bw)
qemu_options+=(size:--disable-vpc)
qemu_options+=(size:--disable-vhdx)
qemu_options+=(size:--disable-hv-balloon)
```

I've disabled the features others have commented about above. @fidencio I added disabling `avx512bw`, unless you see a need to keep it.

ryansavino commented on tools/packaging/scripts/configure-hypervisor.sh:
```diff
@@ -387,6 +387,21 @@ generate_qemu_options() {
+	qemu_options+=(size:--disable-vpc)
+	qemu_options+=(size:--disable-vhdx)
+	qemu_options+=(size:--disable-hv-balloon)
+
```
Right on. Done.

gkurz reviewed (approved):
LGTM. Thanks @ryansavino !
